### ISSUE TITLE
Bug message reported when attempting to support Arrays of elements

### ISSUE THREAD

**Issue by fabriziobertocci** [GaloisInc/daedalus#106]
I have a generic definition of an array of types:
```
-- Arrays of generic type
def ArrayType TypeName len = {
   $$ = Many (len as uint 64) TypeName;
}
```
That is used to create parser like the following:
```
--typedef uint16 ArrayOfShorts[ARR_DIM1][2];
def ArrayOfShorts = ArrayType LE16 8
```
(where `LE16` is defined as: `def LE16 = { @b1 = UInt8; @b2 = UInt8; ^ b2 # b1 }`)

The `ArrayOfShort` is then used in a top-level structure:
```
def Typedefs = {
   ...  sarr = ArrayOfShorts;
}}}
```
When running daedalus I'm getting the following error message:
```
BUG: not enough type arguments for Main.Main. This usually indicates some expression in the program became polymorphic, which can be fixed by adding more type annotations.
CallStack (from HasCallStack):
  error, called at src/Daedalus/Interp.hs:766:7 in daedalus-0.1.0.0-99915268455dde32e69135c9ecdb5545f096a71fc146b48bfe83f417c7705fc9:Daedalus.Interp
```
When running daedalus with `--core` I'm getting:
```
You have encountered a bug in Daedalus's implementation.
*** Please create an issue at https://github.com/GaloisInc/daedalus/issues

%< --------------------------------------------------- 
  Revision:  UNKNOWN
  Branch:    UNKNOWN
  Location:  specialiseOne
  Message:   Specializing a poly function
CallStack (from HasCallStack):
  panic, called at src/Daedalus/Panic.hs:17:9 in daedalus-utils-0.1.0.0-e1469dd5fe02a5d83b73b08f3e3c0baf9aa142674f941d3ea0e91e5b63ed330c:Daedalus.Panic
  panic, called at src/Daedalus/Specialise.hs:159:33 in daedalus-0.1.0.0-99915268455dde32e69135c9ecdb5545f096a71fc146b48bfe83f417c7705fc9:Daedalus.Specialise
%< --------------------------------------------------- 
```

**Comment by yav** [comment]
We really need to report a much better error for this, thanks for reporting it.

TLDR:  You can either remove the `as uint 64` from the definition of `ArrayType` or add a type annotation on the 8, for example `(8 : uint 8)`

The problem, which is reported very poorly here, is that `daedalus` cannot figure out all the types in the program.   This happens because the definition of `ArrayType` has the cast `as uint 64`, which means, roughly, "I am happy with any numeric type, I'll just try to convert it to `uint 64`".    This, combined with the fact that numeric literals are overloaded (i.e., `8` can be used at multiple different types, such as `uint 8`, `uint 32`, or `uint 64`), makes it impossible to know the type of `8` in this example.

**Comment by yav** [comment]
Yes, I fully understand, the little documentation we have as probably also somewhat out of date, so I do appreciate you be willing to experiment with stuff to figure it out!
